src/features/hyperlinks.rs

```diff
@@ -19,26 +19,83 @@ pub fn make_feature() -> Vec<(String, OptionValueFunction)> {
+    // Repeat for matches[1..], and finally store line[pos..].
+    struct HyperlinkCommits<T>(T)
+    where
+        T: Fn(&str) -> String;
+    impl<T: for<'b> Fn(&'b str) -> String> HyperlinkCommits<T> {
+        fn _m(&self, result: &mut String, line: &str, m: &Match, prev_pos: usize) -> usize {
+            result.push_str(&line[prev_pos..m.start()]);
+            let commit = &line[m.start()..m.end()];
+            // Do not link numbers, require at least on non-decimal:
+            if commit.contains(|c| matches!(c, 'a'..='f')) {
+                result.push_str(&self.0(commit));
+            } else {
+                result.push_str(commit);
+            }
+            m.end()
+        }
+        fn with_input(&self, line: &str, m0: &Match, matches123: &mut Matches) -> String {
+            let mut result = String::new();
+            let mut pos = self._m(&mut result, line, m0, 0);
+            // limit number of matches per line, an exhaustive `find_iter` is O(len(line) * len(regex)^2)
+            for m in matches123.take(12) {
+                pos = self._m(&mut result, line, &m, pos);
+            }
+            result.push_str(&line[pos..]);
+            result
+        }
+    }
+
     if let Some(commit_link_format) = &config.hyperlinks_commit_link_format {
-        COMMIT_LINE_REGEX.replace(line, |captures: &Captures| {
-            let prefix = captures.get(1).map(|m| m.as_str()).unwrap_or("");
-            let commit = captures.get(2).map(|m| m.as_str()).unwrap();
-            let suffix = captures.get(3).map(|m| m.as_str()).unwrap_or("");
-            let formatted_commit =
-                format_osc8_hyperlink(&commit_link_format.replace("{commit}", commit), commit);
-            format!("{prefix}{formatted_commit}{suffix}")
-        })
-    } else if let Some(repo) = config.git_config().and_then(GitConfig::get_remote_url) {
-        COMMIT_LINE_REGEX.replace(line, |captures: &Captures| {
-            format_commit_line_captures_with_osc8_commit_hyperlink(captures, &repo)
-        })
-    } else {
-        Cow::from(line)
+        let mut matches = COMMIT_HASH_REGEX.find_iter(line);
+        if let Some(first_match) = matches.next() {
+            let result = HyperlinkCommits(|commit| {
+                format_osc8_hyperlink(&commit_link_format.replace("{commit}", commit), commit)
+            })
+            .with_input(line, &first_match, &mut matches);
+            return Cow::from(result);
+        }
+    } else if let Some(repo) = remote_from_config(&config.git_config()) {
+        let mut matches = COMMIT_HASH_REGEX.find_iter(line);
+        if let Some(first_match) = matches.next() {
+            let result = HyperlinkCommits(|commit| {
+                format!(
+                    "{osc}8;;{url}{st}{commit}{osc}8;;{st}",
+                    url = repo.format_commit_url(commit),
+                    commit = commit,
+                    osc = "\x1b]",
+                    st = "\x1b\\"
```

Comment:
Can be replaced with `format_osc8_hyperlink()`, right?

Reply:
Yes I believe so. I'm not sure why the previous code wasn't already using that. Hopefully just a historical accident rather than a Chesterton's fence.

Reply:
Done (tests agree)!